### PR TITLE
Implement SMS notifications with help of smpp protocol

### PR DESCRIPTION
**Pull request by Abyss777**
- Added `phone` field to `User` model
- Added `sms` field to `Notification` model
- Used velocity engine for SMS templates, because of that have to split templates path on `templates.rootPath`, `mail.templatesPath` and `sms.templatesPath`

`SmppClient` mostly inspired by demo https://github.com/fizzed/cloudhopper-smpp/tree/master/src/test/java/com/cloudhopper/smpp/demo/persist
but very simplified, especially `ReconnectionTask`

Tested on simulator, kannel with opensmppbox and commercial service kindly provided by @jaimzj  

The library uses another log system, because of that some logs go to `wrapper.log`

Will add more comments in code ...

### CODE REVIEW

File: src/org/traccar/api/resource/NotificationResource.java [hunk 1]
```diff
@@ -62,8 +68,10 @@ public Response update(Notification entity) throws SQLException {
 
     @Path("test")
     @POST
-    public Response testMail() throws MessagingException {
+    public Response testMessage() throws MessagingException, RecoverablePduException,
+            UnrecoverablePduException, SmppTimeoutException, SmppChannelException, InterruptedException {
         NotificationMail.sendMailSync(getUserId(), new Event("test", 0), null);
+        NotificationSms.sendSmsSync(getUserId(), new Event("test", 0), null);
```
Comment (Abyss777) [inline src/org/traccar/api/resource/NotificationResource.java]: Test request will try send both

File: src/org/traccar/smpp/EnquireLinkTask.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.traccar.helper.Log;
+
+import com.cloudhopper.smpp.SmppSession;
+import com.cloudhopper.smpp.pdu.EnquireLink;
+import com.cloudhopper.smpp.type.RecoverablePduException;
+import com.cloudhopper.smpp.type.SmppChannelException;
+import com.cloudhopper.smpp.type.SmppTimeoutException;
+import com.cloudhopper.smpp.type.UnrecoverablePduException;
+
+public class EnquireLinkTask implements Runnable {
```
Comment (Abyss777) [inline src/org/traccar/smpp/EnquireLinkTask.java]: It is something like keepalive

File: src/org/traccar/smpp/SmppClient.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    private byte sourceNpi;
+
+    private byte destTon;
+    private byte destNpi;
+
+    public SmppClient() {
+        sessionConfig.setName("Traccar.smppSession");
+        sessionConfig.setInterfaceVersion(
+                (byte) Context.getConfig().getInteger("sms.smpp.version", SmppConstants.VERSION_3_4));
+        sessionConfig.setType(SmppBindType.TRANSCEIVER);
```
Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: We are sending and receiving sms

File: src/org/traccar/smpp/SmppClient.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        sessionConfig.setHost(Context.getConfig().getString("sms.smpp.host", "localhost"));
+        sessionConfig.setPort(Context.getConfig().getInteger("sms.smpp.port", 2775));
+        sessionConfig.setSystemId(Context.getConfig().getString("sms.smpp.username", "user"));
+        sessionConfig.setPassword(Context.getConfig().getString("sms.smpp.password", "password"));
+        sessionConfig.getLoggingOptions().setLogBytes(false);
+        sessionConfig.getLoggingOptions().setLogPdu(Context.getConfig().getBoolean("sms.smpp.logPdu"));
```
Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: This enables logging all PDUs to `wrapper.log`
````
[Reconnection-Traccar.smppSession] INFO com.cloudhopper.smpp.impl.DefaultSmppSession - sync send PDU: (bind_transceiver: 0x00000028 0x00000009 0x00000000 0x00000001) (body: systemId [xxxxxx] password [xxxxxxx] systemType [] interfaceVersion [0x34] addressRange (0x00 0x00 [])) (opts: )
[Traccar.smppSession] INFO com.cloudhopper.smpp.impl.DefaultSmppSession - received PDU: (bind_transceiver_resp: 0x0000001D 0x80000009 0x00000000 0x00000001 result: "OK") (body: systemId [smsplus]) (opts: (sc_interface_version: 0x0210 0x0001 [34]))
[EnquireLink-Traccar.smppSession] INFO com.cloudhopper.smpp.impl.DefaultSmppSession - sync send PDU: (enquire_link: 0x00000010 0x00000015 0x00000000 0x00000002) (body: ) (opts: )
[Traccar.smppSession] INFO com.cloudhopper.smpp.impl.DefaultSmppSession - received PDU: (enquire_link_resp: 0x00000010 0x80000015 0x00000000 0x00000002 result: "OK") (body: ) (opts: )
````

File: src/org/traccar/smpp/SmppClient.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    public synchronized void sendMessageSync(String destAddress, String message) throws RecoverablePduException,
+            UnrecoverablePduException, SmppTimeoutException, SmppChannelException, InterruptedException {
+        if (getSession() != null && getSession().isBound()) {
+            byte[] textBytes = CharsetUtil.encode(message, charsetName);
+
+            SubmitSm submit = new SubmitSm();
+            submit.setSourceAddress(new Address(sourceTon, sourceNpi, sourceAddress));
+            submit.setDestAddress(new Address(destTon, destNpi, destAddress));
+            submit.setRegisteredDelivery(SmppConstants.REGISTERED_DELIVERY_SMSC_RECEIPT_REQUESTED);
```
Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: Always requesting delivery report

File: src/org/traccar/smpp/SmppClient.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            submit.setDataCoding((byte) dataCoding);
+            submit.setShortMessage(textBytes);
+            SubmitSmResp submitResponce = getSession().submit(submit, submitTimeout);
```
Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: Submitting message in synchronous way. It is enough on the current stage.

File: src/org/traccar/smpp/SmppClient.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            Log.debug("SMS submited, msg_id: " + submitResponce.getMessageId());
```
Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: We can trace message deliver by id in future, now just log.

File: src/org/traccar/smpp/SmppClient.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            throw new SmppChannelException("Smpp session is not connected");
+        }
+    }
+
+    public void sendMessageAsync(final String destAddress, final String message) {
+        executorService.execute(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    sendMessageSync(destAddress, message);
```
Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: Send messages in separate threads

File: src/org/traccar/smpp/SmppClient.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        });
+
+        scheduleReconnect();
+    }
+
+    public synchronized SmppSession getSession() {
+        return smppSession;
+    }
+
+    public String mapDataCodingToCharset(byte dataCoding) {
```
Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: This function is not full, there is a problem to map `dataCoding` to charsets https://github.com/twitter/cloudhopper-commons/issues/7

File: src/org/traccar/smpp/ClientSmppSessionHandler.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    public void firePduRequestExpired(PduRequest pduRequest) {
+        Log.warning("PDU request expired: " + pduRequest);
+    }
+
+    @Override
+    public PduResponse firePduRequestReceived(PduRequest request) {
+        PduResponse response = null;
+        try {
+            if (request instanceof DeliverSm) {
+                Log.debug("Message Received: "
```
Comment (Abyss777) [inline src/org/traccar/smpp/ClientSmppSessionHandler.java]: Here we are logging delivers and incoming sms
````
2017-02-21 12:31:19 DEBUG: Message Received: id:56dfb7e2-a336-4432-8b2e-caaa0f897662 sub:001 dlvrd:001 submit date:170221130039 done date:170221130111 stat:DELIVRD err:000 text:, Source Address: 7922xxxxxx
````

**Comment by jaimzj** [comment]
A small feature request, I think it would appropriate to have this so that Admin>Managers can have better control over the cost incurred by them, As SMS is chargeable and the SMPP settings are configurable only on a server level and not user level. I feel it would be of great use if there is a per user SMS limit count, which can be almost similar to device limit implementation in the User ROLE management module? 
-Except that maybe SMS count can be topped up by the superior role.
- Admin should be able to provide lets say 100 SMS to Manager and Manager int-turn can distribute that divide into his number of users and allocate.

Maybe there is a better solution to this, but this just a suggestion :) please correct me.

**Comment by tananaev** [comment]
@jaimzj, please create a separate thread for your feature request. I don't think it's a good idea to implement it as a part of this pull request.

File: src/org/traccar/smpp/ClientSmppSessionHandler.java (hunk 10 above)
Comment (tananaev) [inline src/org/traccar/smpp/ClientSmppSessionHandler.java]: I think we can simplify this log. I don't think we need all of this information in the log.

File: src/org/traccar/smpp/EnquireLinkTask.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public void run() {
+        SmppSession smppSession = smppClient.getSession();
+        if (smppSession != null && smppSession.isBound()) {
+            try {
+                smppSession.enquireLink(new EnquireLink(), enquireLinkTimeout);
+            } catch (SmppTimeoutException | SmppChannelException
+                    | RecoverablePduException | UnrecoverablePduException error) {
+                Log.warning("Enquire link failed, executing reconnect: " + error);
```
Comment (tananaev) [inline src/org/traccar/smpp/EnquireLinkTask.java]: Exception can be included as a separate parameter.

File: src/org/traccar/smpp/EnquireLinkTask.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                smppClient.reconnect();
+            } catch (InterruptedException error) {
+                Log.info("Enquire link interrupted, probably killed by reconnecting");
+            }
+        } else {
+            Log.error("Enquire link running while session is not connected");
```
Comment (tananaev) [inline src/org/traccar/smpp/EnquireLinkTask.java]: Error is reserved for fatal exceptions. Please use warning.

File: src/org/traccar/smpp/SmppClient.java (hunk 4 above)
Comment (tananaev) [inline src/org/traccar/smpp/SmppClient.java]: It would be nice to find a way to log it into server log, but not critical at the moment.

Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: Agree, but cloudhopper uses slf4j, traccar uses log4j, there is only log4j -> slf4j but we need opposite.
I'll be really appreciate if you point me a solution.

Comment (tananaev) [inline src/org/traccar/smpp/SmppClient.java]: Let's leave it as it is for now.

File: src/org/traccar/smpp/ClientSmppSessionHandler.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        Log.warning("PDU request expired: " + pduRequest);
+    }
+
+    @Override
+    public PduResponse firePduRequestReceived(PduRequest request) {
+        PduResponse response = null;
+        try {
+            if (request instanceof DeliverSm) {
+                if (request.getOptionalParameters() != null) {
+                    Log.debug("Message Delivered: "
```
Comment (Abyss777) [inline src/org/traccar/smpp/ClientSmppSessionHandler.java]: Now delivery reports look like this:
````
2017-02-22 16:13:50 DEBUG: Message Delivered: 9433835d-cbe4-4838-b13e-34f31dacba70, State: 2
````

Where `State` is https://github.com/fizzed/cloudhopper-smpp/blob/master/src/main/java/com/cloudhopper/smpp/SmppConstants.java#L288

File: src/org/traccar/smpp/ClientSmppSessionHandler.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+                            + request.getOptionalParameter(SmppConstants.TAG_RECEIPTED_MSG_ID).getValueAsString()
+                            + ", State: "
+                            + request.getOptionalParameter(SmppConstants.TAG_MSG_STATE).getValueAsByte());
+                } else {
+                    Log.debug("Message Received: "
```
Comment (Abyss777) [inline src/org/traccar/smpp/ClientSmppSessionHandler.java]: Incoming message look like this:
````
2017-02-22 16:07:49 DEBUG: Message Received: Incoming message, Source Address: +7922xxxxxxxx
````

**Comment by Abyss777** [comment]
Fixed logging

File: src/org/traccar/notification/NotificationFormatter.java [hunk 15]
```diff
@@ -47,18 +48,42 @@ public static MailMessage formatMessage(long userId, Event event, Position posit
...
         }
 
         StringWriter writer = new StringWriter();
         template.merge(velocityContext, writer);
         String subject = (String) velocityContext.get("subject");
         return new MailMessage(subject, writer.toString());
     }
+
+    public static String formatSmsMessage(long userId, Event event, Position position) {
+        VelocityContext velocityContext = prepareContext(userId, event, position);
```
Comment (tananaev) [inline src/org/traccar/notification/NotificationFormatter.java]: I think we can share most of the code in this method.

File: src/org/traccar/smpp/SmppClient.java (hunk 5 above)
Comment (tananaev) [inline src/org/traccar/smpp/SmppClient.java]: Do we actually want to request it? We don't handle it, right?

Comment (Abyss777) [inline src/org/traccar/smpp/SmppClient.java]: Not really for now. Right, just log.
I've made it more for testing. Will remove it now. We can revert it in future, when will control delivery.

**Comment by Abyss777** [comment]
It is not convenient to share code in `prepareContext` because we need `subject` in one case. I've combined it in other function.
Also removed delivery receipt request.

File: src/org/traccar/notification/NotificationFormatter.java [hunk 16]
```diff
@@ -47,18 +48,39 @@ public static MailMessage formatMessage(long userId, Event event, Position posit
             velocityContext.put("geofence", Context.getGeofenceManager().getGeofence(event.getGeofenceId()));
         }
         velocityContext.put("webUrl", Context.getVelocityEngine().getProperty("web.url"));
+        return velocityContext;
+    }
 
-        Template template = null;
+    private static Template getTemplate(Event event, boolean mail) {
```
Comment (tananaev) [inline src/org/traccar/notification/NotificationFormatter.java]: I think it would make more sense to just pass the path string into this method.

**Comment by Abyss777** [comment]
Totally agree. Looks much better.
Done.

**Comment by tananaev** [comment]
Merged, thanks. Can you please update configuration documentation and ideally prepare some manual on how to configure SMS (and maybe other) notifications.

**Comment by Abyss777** [comment]
Should I edit https://www.traccar.org/configuration-file/ directly or provide you a draft somehow?

**Comment by tananaev** [comment]
You can change it directly if you want. Just let me know which parameters you've changed/added, so I can review it.

**Pull request merged**